# Telecom Churn Case Study Hackathon

# 1. Problem Statement

In the telecom industry, customers are able to choose from multiple service providers and actively switch from one operator to another. In this highly competitive market, the telecommunications industry experiences an average of 15-25% annual churn rate. Given the fact that it costs 5-10 times more to acquire a new customer than to retain an existing one, customer retention has now become even more important than customer acquisition.

For many incumbent operators, retaining high profitable customers is the number one business
goal. To reduce customer churn, telecom companies need to predict which customers are at high risk of churn. In this project, you will analyze customer-level data of a leading telecom firm, build predictive models to identify customers at high risk of churn.

In this competition, your goal is to build a machine learning model that is able to predict churning customers based on the features provided for their usage.

# 2. Business Objectives

Predict churning customers for a Telecom company based on temporal behaviour

# 3. IMPORT LIBRARIES AND DATASETS

In [1]:
# import package
import numpy as np                                                 # Implemennts milti-dimensional array and matrices
import pandas as pd                                                # For data manipulation and analysis
import matplotlib.pyplot as plt                                    # Plotting library for Python programming language and it's numerical mathematics extension NumPy
import seaborn as sns                                              # Provides a high level interface for drawing attractive and informative statistical graphics

pd.set_option('display.max_columns', None)
%matplotlib inline
sns.set()

In [2]:
import warnings
warnings.filterwarnings('ignore')

## 3.1. Data Loading and Description

- **train.csv** - Contains 172 columns. The primary key that represents each customer is id. The target variable that you need to predict is churn_probability which contains a value of 0 or 1. This data is what you are going to use for EDA, cleaning, feature engineering, model building, model evaluation, model selection, and finally model training.
- **test.csv** - Contains 171 columns, doesn't contain the target variable churn_probability. You will apply all the necessary preprocessing steps to get this data into the right format and then use the model trained using the train.csv file to make predictions with this. This is unseen data! Only the competition hosts know the actual values of the target feature for this data and therefore your submissions will be evaluated on how well your model does with this dataset.
- **sample.csv** - This contains the format in which you need to submit the solutions to Kaggle. The id column in this dataset exactly the same as the id column in test.csv. You will make your predictions on the test.csv data and store them in a submission file that has the same format as this file. Check the Overview>Evaluation tab for mode details.
- **Data Dictionary** - This contains the definitions for the various acronyms that you will need to understand each variable. For example, the variable total_og_mou_7, contains the acronyms total, og, mou, and 7, which can be translated as the total outgoing minutes of voice calls made by the user in month of July.

To gain insights from data we must look into each aspect of it very carefully. We will start with observing few rows and columns of data both from the starting and from the end.

Let us check the basic information of the dataset. The very basic information to know is the dimension of the dataset – rows and columns – that’s what we find out with the method shape.

In [3]:
df = pd.read_csv("/kaggle/input/telecom-churn-case-study-hackathon-c42/train.csv")

In [4]:
df.sample(5)

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
36257  36257        109             0.0             0.0             0.0   
38676  38676        109             0.0             0.0             0.0   
4697    4697        109             0.0             0.0             0.0   
29766  29766        109             0.0             0.0             0.0   
27869  27869        109             0.0             0.0             0.0   

      last_date_of_month_6 last_date_of_month_7 last_date_of_month_8   arpu_6  \
36257            6/30/2014            7/31/2014            8/31/2014  171.471   
38676            6/30/2014            7/31/2014            8/31/2014  207.444   
4697             6/30/2014            7/31/2014            8/31/2014  355.440   
29766            6/30/2014            7/31/2014            8/31/2014  337.412   
27869            6/30/2014            7/31/2014            8/31/2014  336.682   

        arpu_7   arpu_8  onnet_mou_6  onnet_mou_7  onnet_mou_8  offnet_mou_6  \
36257  539.643    0.000       192.99       635.93         0.00         36.43   
38676  112.758  360.807       102.66       105.54       107.98         26.39   
4697   433.152  444.608        19.93        19.43         8.89         56.69   
29766  278.469    0.000       368.78       282.19          NaN        208.99   
27869  281.959  349.560        23.48        32.38        56.88         94.08   

       offnet_mou_7  offnet_mou_8  roam_ic_mou_6  roam_ic_mou_7  \
36257        278.14          0.00            0.0            0.0   
38676         21.86         12.84            0.0            0.0   
4697          85.84         82.54            0.0            0.0   
29766        162.73           NaN            0.0            0.0   
27869         62.74         95.86            0.0            0.0   

       roam_ic_mou_8  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  \
36257            0.0            0.0            0.0            0.0   
38676            0.0            0.0            0.0            0.0   
4697             0.0            0.0            0.0            0.0   
29766            NaN            0.0            0.0            NaN   
27869            0.0            0.0            0.0            0.0   

       loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2m_mou_6  \
36257              6.64              5.04              0.00             21.86   
38676              3.78              0.00              1.45             20.74   
4697               0.00              0.00              0.00              2.33   
29766              7.26             10.11               NaN              3.04   
27869             23.48             31.73             56.88             84.29   

       loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2f_mou_6  loc_og_t2f_mou_7  \
36257             54.29              0.00              3.10              0.00   
38676             18.14             11.56              1.41              0.00   
4697               0.95              1.93              0.00              0.00   
29766              4.00               NaN              0.00              0.00   
27869             51.74             85.81              0.00              0.95   

       loc_og_t2f_mou_8  loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  \
36257               0.0              0.00              0.00              0.00   
38676               0.0              0.00              0.00              0.00   
4697                0.0              0.78             36.08              3.23   
29766               NaN              0.00              0.00               NaN   
27869               0.0              0.00              1.33              0.00   

       loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  std_og_t2t_mou_6  \
36257         31.61         59.34          0.00            186.34   
38676         25.94         18.14         13.01             98.88   
4697           2.33          0.95          1.93             19.93   
29766         10.31         14.11   

## 3.2. Data Understanding

In [5]:
df.shape

(69999, 172)

The dataset has __69999 rows and 172 columns.__

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69999 entries, 0 to 69998
Columns: 172 entries, id to churn_probability
dtypes: float64(135), int64(28), object(9)
memory usage: 91.9+ MB


In [7]:
# Get info from both data set to understand data in depths 
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69999 entries, 0 to 69998
Data columns (total 172 columns):
 #    Column                    Dtype  
---   ------                    -----  
 0    id                        int64  
 1    circle_id                 int64  
 2    loc_og_t2o_mou            float64
 3    std_og_t2o_mou            float64
 4    loc_ic_t2o_mou            float64
 5    last_date_of_month_6      object 
 6    last_date_of_month_7      object 
 7    last_date_of_month_8      object 
 8    arpu_6                    float64
 9    arpu_7                    float64
 10   arpu_8                    float64
 11   onnet_mou_6               float64
 12   onnet_mou_7               float64
 13   onnet_mou_8               float64
 14   offnet_mou_6              float64
 15   offnet_mou_7              float64
 16   offnet_mou_8              float64
 17   roam_ic_mou_6             float64
 18   roam_ic_mou_7             float64
 19   roam_ic_mou_8             float64
 20   roam

We have following 9 Columns as object type
- last_date_of_month_6
- last_date_of_month_7
- last_date_of_month_8
- date_of_last_rech_6
- date_of_last_rech_7
- date_of_last_rech_8
- date_of_last_rech_data_6
- date_of_last_rech_data_7
- date_of_last_rech_data_8

In [8]:
df.describe()

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
count  69999.000000    69999.0         69297.0         69297.0   
mean   34999.000000      109.0             0.0             0.0   
std    20207.115084        0.0             0.0             0.0   
min        0.000000      109.0             0.0             0.0   
25%    17499.500000      109.0             0.0             0.0   
50%    34999.000000      109.0             0.0             0.0   
75%    52498.500000      109.0             0.0             0.0   
max    69998.000000      109.0             0.0             0.0   

       loc_ic_t2o_mou        arpu_6        arpu_7        arpu_8   onnet_mou_6  \
count         69297.0  69999.000000  69999.000000  69999.000000  67231.000000   
mean              0.0    283.134365    278.185912    278.858826    133.153275   
std               0.0    334.213918    344.366927    351.924315    299.963093   
min               0.0  -2258.709000  -1289.715000   -945.808000      0.000000   
25%               0.0     93.581000     86.714000     84.095000      7.410000   
50%               0.0    197.484000    191.588000    192.234000     34.110000   
75%               0.0    370.791000    365.369500    369.909000    119.390000   
max               0.0  27731.088000  35145.834000  33543.624000   7376.710000   

        onnet_mou_7   onnet_mou_8  offnet_mou_6  offnet_mou_7  offnet_mou_8  \
count  67312.000000  66296.000000  67231.000000  67312.000000  66296.000000   
mean     133.894438    132.978257    198.874771    197.153383    196.543577   
std      311.277193    311.896596    316.818355    322.482226    324.089234   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        6.675000      6.410000     34.860000     32.240000     31.575000   
50%       32.280000     32.100000     96.480000     91.885000     91.800000   
75%      115.837500    115.060000    232.990000    227.630000    229.345000   
max     8157.780000  10752.560000   8362.360000   7043.980000  14007.340000   

       roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  roam_og_mou_6  \
count   67231.000000   67312.000000   66296.000000   67231.000000   
mean        9.765435       7.014568       7.004892      14.186457   
std        57.374429      55.960985      53.408135      73.469261   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max      2850.980000    4155.830000    4169.810000    3775.110000   

       roam_og_mou_7  roam_og_mou_8  loc_og_t2t_mou_6  loc_og_t2t_mou_7  \
count   67312.000000   66296.000000      67231.000000      67312.000000   
mean        9.842191       9.771783         46.904854         46.166503   
std        58.511894      64.618388        150.971758        154.739002   
min         0.000000       0.000000          0.000000          0.000000   
25%         0.000000       0.000000          1.660000          1.650000   
50%         0.000000       0.000000         11.910000         11.580000   
75%         0.000000       0.000000         40.740000         39.760000   
max      2812.040000    5337.040000       6431.330000       7400.660000   

       loc_og_t2t_mou_8  loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  \
count      66296.000000      67231.000000      67312.000000      66296.000000   
mean          45.686109         93.238231         90.799240         91.121447   
std          153.716880        162.046699        153.852597        152.997805   
min            0.000000          0.000000          0.000000          0.000000   
25%            1.610000          9.920000         10.090000          9.830000   
50%           11.740000         41.030000         40.170000         40.350000   
75%           39.895000        110.430000        107.540000        109.245000   
max        10752.560000       4696

- The dataset comprises of 69999 observations of 172 columns. Above is a table showing names of all the columns and their description.
- There are no categorical value present in dataset, rest of columns are integer/float type.
- Also few columns have object type as date value which we can use to create a derive data from that

# 4. Data Engineering 

Below are the points that need to be addressed in this step:
1. Duplicates
2. Data validity check
3. Missing values
4. Outliers

## 4.1. Data Duplicates

The data may contain duplicate entries, which need to be removed in most of the applications.

In [9]:
duplicate_df = df[df.duplicated()]
duplicate_df.shape

(0, 172)

There are duplicates entrires present in traing dataset

## 4.2. Data validity check

Based on the features, need to validate the data.

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69999 entries, 0 to 69998
Columns: 172 entries, id to churn_probability
dtypes: float64(135), int64(28), object(9)
memory usage: 91.9+ MB


We have arount 172 features. However there are 9 Columns as object type & might not useful for our analysis.

- last_date_of_month_6
- last_date_of_month_7
- last_date_of_month_8
- date_of_last_rech_6
- date_of_last_rech_7
- date_of_last_rech_8
- date_of_last_rech_data_6
- date_of_last_rech_data_7
- date_of_last_rech_data_8

## 4.3. Missing values

- Based on the business, a missing value imputation needs to be chosen.
- If the data is normally distributed, mean imputation can be performed. If the mean and median difference is huge, median imputation is preferred.

In [11]:
# Checking more meta information about dataframe i.e. how many non-null, uniquie value.

def metadataInfo(dataframe) : 
    return pd.DataFrame({
                'Datatype' : dataframe.dtypes.astype(str), 
                'Non_Null_Count': dataframe.count(axis = 0).astype(int), 
                'Null_Count': dataframe.isnull().sum().astype(int), 
                'Null_Percentage': round(dataframe.isnull().sum()/len(dataframe) * 100 , 2), 
                'Unique_Values_Count': dataframe.nunique().astype(int) 
                 }).sort_values(by='Null_Percentage', ascending=False)

In [12]:
new_df = metadataInfo(df)
print(new_df.to_string())

                         Datatype  Non_Null_Count  Null_Count  Null_Percentage  Unique_Values_Count
arpu_3g_6                 float64           17568       52431            74.90                 5506
count_rech_2g_6           float64           17568       52431            74.90                   30
night_pck_user_6          float64           17568       52431            74.90                    2
arpu_2g_6                 float64           17568       52431            74.90                 5389
date_of_last_rech_data_6   object           17568       52431            74.90                   30
total_rech_data_6         float64           17568       52431            74.90                   36
av_rech_amt_data_6        float64           17568       52431            74.90                  792
max_rech_data_6           float64           17568       52431            74.90                   47
count_rech_3g_6           float64           17568       52431            74.90                   23


There are so many __Null/NA__ value present in dataset which varies from 1% to 74%.
- Columns which have more that 60% __NULL/NA__ value, we will drop those since we might not more information from those colum
- However columns which have minimum Null value present and if the data is normally distributed in those, then mean imputation can be performed. If the mean and median difference is huge, median imputation is preferred on numerical data.
- Since we don't have categorical data present, no need to perform Mode imputation on categorical data.

## 4.4. Outliers

- Outliers can be valid outlier or data entry mistake. 
- Need to understand what kind of outlier the data contains. Based on the outlier type, the issue needs to be addressed.

# 5. Exploratory Data Analysis 

This step gives more insights into the data and how the data is related to the target variable. This step involves mainly:
1. Uni-variate analysis
2. Bi-variate/Multi-variate analysis
3. Pivots
4. Visualization and Data Insights

# 6. Feature Engineering and Selection

- Feature engineering involves identifying the right drivers/features that affect the target variable and also deriving the new features based on existing features.
- Based on feature importance, some of the features can be removed from the data, which helps in reducing data size. 

# 7. Model Building

Model building is the process of developing a probabilistic model that best describes the relationship between independent and dependent variables. Various ML models are to be built based on the problem statement.

Below are the topics that can be covered in this step :
1. Creating dummy variables
2. Over Sampling and Under Sampling (if the data is imbalanced)
3. Split the data into train and test

# 8. Model Validation & Evaluation

This step covers
1. Testing the model
2. Tuning the model
3. Cross-validation
4. Model evaluation metrics trade-off
5. Model Underfitting/Overfitting

## 8.1. Testing the model :

Run the model on test data and evaluate the model performance using the correct metric based on the business use case.

## 8.2. Tuning the model :

Model tuning involves, improving model performance by iterating the parameter values during model building. After fine-tuning, the model needs to re-build.

## 8.3. Cross-validation :

Cross-validation is used to evaluate how the model will perform on an independent test dataset. Some of the cross-validation techniques are :
1. K- fold cross-validation

## 8.4. Model Evaluation metrics trade-off :

Trade-offs always help us to find the sweet spot or the middle ground. Machine learning mostly deals with two tradeoffs :
1. Bias-variance tradeoff
2. Precision-Recall tradeoff

## 8.5. Model Underfitting and Overfitting :

- Overfitting is the case, where the model is trying to capture all the patterns in the training data and failing to perform in the test data.
- Underfitting is the case, where the model is not learning properly in the training data and also not performing well in test data.

# 9. Model Predict

# 10. Conclusion 